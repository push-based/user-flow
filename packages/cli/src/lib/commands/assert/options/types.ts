import { SharedFlagsSettings } from 'lighthouse/types/lhr/settings.js';

// @TODO this type has overlap with the one in rc-json.ts we should fix that and only have one
export type AssertRcOptions = {
  budgetPath?: string,
  budgets?: SharedFlagsSettings['budgets']
}
export type AssertArgvOptions = AssertRcOptions;

