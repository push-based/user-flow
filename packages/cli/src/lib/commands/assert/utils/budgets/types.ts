import { default as _Budget } from 'lighthouse/types/lhr/budget';

export interface Budget {
  resourceType: _Budget.ResourceType,
  sizeOverBudget: number;
  transferSize: number;
  size: number;
  countOverBudget: string;
  requestCount: number;
}
export type BudgetDetails = { items: (Budget)[] }

