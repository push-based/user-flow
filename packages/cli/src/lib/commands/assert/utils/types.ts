/* eslint-disable */
/**
 * @license Copyright 2019 Google Inc. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */

type StrictOmit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;


export type AssertionAggregationMethod =
  | 'median'
  | 'optimistic'
  | 'pessimistic'
  | 'median-run';

export type AssertionFailureLevel = 'off' | 'warn' | 'error';

export interface AssertionOptions {
  aggregationMethod?: AssertionAggregationMethod;
  minScore?: number;
  maxLength?: number;
  maxNumericValue?: number;
}

export interface Assertions {
  [auditId: string]: AssertionFailureLevel | [AssertionFailureLevel, AssertionOptions];
}

export interface BaseOptions {
  matchingUrlPattern?: string;
  aggregationMethod?: AssertionAggregationMethod;
  preset?: 'lighthouse:all' | 'lighthouse:recommended';
  assertions?: Assertions;
}

export interface Options extends BaseOptions {
  includePassedAssertions?: boolean;
  budgetsFile?: string;
  assertMatrix?: BaseOptions[];
}


export type AssertionType =
  | keyof StrictOmit<AssertionOptions, 'aggregationMethod'>
  | 'auditRan';

export interface AssertionResult {
  url: string;
  name: keyof Omit<AssertionOptions, 'aggregationMethod'> | 'auditRan';
  operator: string;
  expected: number;
  actual: number;
  values: number[];
  passed: boolean;
  level?: AssertionFailureLevel | {};
  auditId?: string;
  auditProperty?: string;
  auditTitle?: string;
  auditDocumentationLink?: string;
  message?: string;
}

export type AssertionResultNoURL = StrictOmit<AssertionResult, 'url'>;
