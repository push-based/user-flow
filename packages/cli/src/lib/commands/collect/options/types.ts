import { Param as Open } from './open.model';
import { Param as UfPath } from './ufPath.model';
import { Param as OutPath } from './outPath.model';
import { Param as Url } from './url.model';
import { Param as ServeCommand } from './serveCommand.model';
import { Param as AwaitServeStdout } from './awaitServeStdout.model';
import { AssertOptions } from '../../assert/options/types';
import { YargsArgvOptionFromParamsOptions } from '../../../core/yargs/types';

export type CollectYargsOptions = Open & UfPath & OutPath & Url & ServeCommand & AwaitServeStdout & AssertOptions;
export type CollectArgvOptions = YargsArgvOptionFromParamsOptions<CollectYargsOptions>;
