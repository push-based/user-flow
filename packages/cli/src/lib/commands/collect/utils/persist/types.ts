import { CollectArgvOptions, PersistArgvOptions } from '../../options/types';

export type PersistFlowOptions = Pick<
  PersistArgvOptions,
  'outPath' | 'format'
> &
  Pick<CollectArgvOptions, 'url'>;
