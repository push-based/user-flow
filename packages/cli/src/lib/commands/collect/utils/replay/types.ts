import { UserFlow, Step } from '@puppeteer/replay';
import { Modify } from '../../../../core/types';

/**
 *  'navigation' is already covered by `@puppeteer/replay`
 */
export type MeasureModes =
  | 'navigate'
  | 'snapshot'
  | 'startTimespan'
  | 'endTimespan';

/*
// Consider modify the Step type
  | Modify<Step, {
  type: MeasureModes,
}>;*/
export type MeasurementStep = {
  type: MeasureModes;
  stepOptions?: { stepName?: string };
  url?: string;
};

export type UserFlowRecordingStep = MeasurementStep | Step;

export type UserFlowReportJson = Modify<
  UserFlow,
  {
    steps: UserFlowRecordingStep[];
  }
>;

export type ReadFileExtTypes = { json: {}; html: string; text: string };
export type ReadFileConfig = { fail?: boolean; ext?: keyof ReadFileExtTypes };
