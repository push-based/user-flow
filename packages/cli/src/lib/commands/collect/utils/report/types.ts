import Budget from 'lighthouse/types/lhr/budget';
import Config from 'lighthouse/types/config';
import { CLI_MODES } from '../../../../global/cli-mode';
import { PickOne } from '../../../../core/types';
import FlowResult from 'lighthouse/types/lhr/flow';

type OverBudget = { overBudget: number };
type BudgetAssertion = (Budget.ResourceBudget & OverBudget | Budget.TimingBudget & OverBudget)[];

type UfrSlice = PickOne<FlowResult>;
type LhrSlice = PickOne<FlowResult.Step['lhr']>;
/**
 * Plucks key value from oroginal LH report
 * @example
 *
 * const t: GatherModeSlice = {gatherMode: 'navigation'};
 * const f1: GatherModeSlice = {gatherMode: 'timespan'};
 * const f2: GatherModeSlice = {gatherMode: 'snapshot'};
 */
type LhrGatherModeSlice = LhrSlice & { gatherMode: FlowResult.Step['lhr']['gatherMode'] };
type UfrNameSlice = UfrSlice & { name: string };


/**
 * This type is the result of `calculateCategoryFraction` https://github.com/GoogleChrome/lighthouse/blob/master/core/util.cjs#L540.
 * As there is no typing present ATM we maintain our own.
 */
export type FractionResults = {
  numPassed: number;
  numPassableAudits: number;
  numInformative: number;
  totalWeight: number;
}

export type ReducedFlowStep =
  // gatherMode
  LhrGatherModeSlice &
  {
    name: string;
    fetchTime: string;
    results: ReducedFlowStepResult;
    resultsPerformanceBudget: any,
    baseline?: ReducedFlowStepResult;
  };

export type ReducedReport = {
  cliMode?: CLI_MODES;
  dryRun?: boolean;
  headless?: boolean;
  name: string;
  fetchTime: string;
  steps: ReducedFlowStep[];
  assertions?: {
    lhBudgetAssertion: BudgetAssertion,
    baselineAssertion: BudgetAssertion,
  }
  config?: Config.Json & { baseline?: any };
}

export type ReducedFlowStepResult = Record<string, number | FractionResults>;
