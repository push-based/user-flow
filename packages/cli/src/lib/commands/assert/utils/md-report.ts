import {FractionResults, GatherMode, ReducedFlowStep, ReducedReport} from '../../collect/utils/report/types';
import {enrichReducedReportWithBaseline} from '../../collect/utils/report/utils';
import {Alignment, table} from '../../../core/md/table';
import {style} from '../../../core/md/font-style';
import {headline, Hierarchy} from '../../../core/md/headline';
import {NEW_LINE} from '../../../core/md/constants';
import {details} from '../../../core/md/details';
import Budget from "lighthouse/types/lhr/budget";
import TimingBudget = Budget.TimingBudget;
import Details from "lighthouse/types/lhr/audit-details";
import Table = Details.Table;

const budgetsSymbol = 'ðŸ”’'

export function generateMdReport(flowResult: ReducedReport): string {
  const name = flowResult.name;
  const dateTime = `Date/Time: ${style(new Date().toISOString().replace('T', ' ').split('.')[0].slice(0, -3))}  `;
  const stepsTable = getStepsTable(flowResult);

  let md = `${headline(name)}${NEW_LINE}
${dateTime}${NEW_LINE}
${stepsTable}${NEW_LINE}
`;

  const budgetsTable = getBudgetTable(flowResult);
  if (budgetsTable !== '') {
    md += details(`${budgetsSymbol} Budgets`, budgetsTable) + NEW_LINE;
  }


  return md;
}

/**
 * ## Navigation report (127.0.0.1/)
 *
 * | Resource Type | Requests | Transfer Size | Over Budget |
 * | :-----------: | :------: | :-----------: | :---------: |
 * |     Total     |  205774  |     total     |   204750    |
 * |    Script     |  101958  |    script     |      -      |
 * |  Third-party  |  97308   |  third-party  |      -      |
 *
 * |         Metric         | Measurement | Over Budget |
 * | :--------------------: | :---------: | :---------: |
 * | First Meaningful Paint |    2043     |     43      |
 * |  Time to Interactive   |    4745     |      -      |
 *
 */

export function getBudgetTable(reducedReport: ReducedReport, options: { heading: Hierarchy } = {heading: 3}): string {
  const performanceBudgets = reducedReport.steps
    .filter(({resourceCountsBudget, resourceSizesBudget, timingsBudget}) => resourceCountsBudget || resourceSizesBudget || timingsBudget)
    .map(({name, resourceCountsBudget, resourceSizesBudget, timingsBudget}) => ({
        name,
        resultsSizeBudget: getResourceSizes(resourceSizesBudget),
        resultsCountBudget: getResourceCounts(resourceCountsBudget),
        resultsTimingBudget: getTimings(timingsBudget)
      })
    );
  return performanceBudgets.length ? performanceBudgets.map(b => {
    let md = headline(b.name, options.heading) + NEW_LINE + NEW_LINE;
    if (b.resultsSizeBudget !== undefined) {
      md += style('Resource Size Budget') + NEW_LINE + NEW_LINE;
      md += table(b.resultsSizeBudget) + NEW_LINE + NEW_LINE
    }
    if (b.resultsCountBudget !== undefined) {
      md += style('Resource Count Budget') + NEW_LINE + NEW_LINE;
      md += table(b.resultsCountBudget) + NEW_LINE + NEW_LINE
    }
    if (b.resultsTimingBudget !== undefined) {
      md += style('Timing Budget') + NEW_LINE + NEW_LINE;
      md += table(b.resultsTimingBudget) + NEW_LINE + NEW_LINE
    }
    return md;
  }).join(NEW_LINE) : '';
}


function getTimings(resultsTimingBudget: Table | undefined): undefined | string[][] {
  if (resultsTimingBudget === undefined) {
    return undefined
  } else {
    return [
      // [ {text: string, ...props } ]
      resultsTimingBudget.headings.map((h: any) => h.text as string),
      ...resultsTimingBudget.items.map(({label, measurement, overBudget}: any) => {
        const row = [];
        if(label === 'Cumulative Layout Shift') {
          return [label + '',
            measurement === undefined ? '-' : parseFloat(measurement).toFixed(2) + '---'+measurement,
            overBudget === undefined ? '-' : parseFloat(overBudget).toFixed(2)+ '---'+overBudget,
          ]
        } else {
          return [label + '',
            measurement === undefined ? '-' : measurement + ' ms',
            overBudget === undefined ? '-' : overBudget + ' ms'
          ]
        }
      })
    ]
  }
};

function getResourceSizes(resourceSizesBudget: Table | undefined): undefined | string[][] {
  if (resourceSizesBudget === undefined) {
    return undefined
  } else {
    return [
      resourceSizesBudget.headings
        .filter((i: any) => ['label', 'transferSize', 'sizeOverBudget'].includes(i.key))
        .map((h: any) => h.text as string),
      ...resourceSizesBudget.items
        .map(
          ({label, transferSize, sizeOverBudget}: any) =>
            [
              label+'',
              formatBytes(transferSize),
              sizeOverBudget ? formatBytes(sizeOverBudget) : '-'
            ])
    ]
  }
}

function getResourceCounts(resourceCountsBudget: Table | undefined): undefined | string[][] {
  if (resourceCountsBudget === undefined) {
    return undefined
  } else {
    return [
      resourceCountsBudget.headings
        .filter((i: any) => ['label', 'requestCount', 'countOverBudget'].includes(i.key))
        .map((h: any) => h.key === 'countOverBudget' ? 'Over Budget' : h.text as string),
      ...resourceCountsBudget.items
        .map(
          ({label, requestCount, countOverBudget}: any) =>
            [
              label+'',
              requestCount+'',
              countOverBudget ? countOverBudget : '-'
            ])
    ]
  }
}



/**
 * | Step Name       | Gather Mode |Performance | Accessibility | BestPractices | Seo  | PWA |
 * | --------------- | ----------- | ------------- | ------------- | ---- | --- |
 * |  Nav1           |  99         | 50            | 100           | 98   |  -  |
 * |  Snap   1       |  3/3        | 22/5          | 5/2           | 7/10 |  -  |
 * |  TimeSpan 1     |  10/11      | -             | 4/7           | 7/10 |  -  |
 */
export function getStepsTable(reducedReport: ReducedReport, baselineResults?: any): string {
  const reportCategories = Object.keys(reducedReport.steps[0].results);
  const tableStepsArr = formatStepsForMdTable(reportCategories, reducedReport, baselineResults);
  const alignOptions = headerAlignment(reportCategories);
  const tableArr = extractTableArr(reportCategories, tableStepsArr);
  return table(tableArr, alignOptions);
}

function formatStepsForMdTable(reportCategories: string[], reducedReport: ReducedReport, baselineResults?: any): string[][] {
  if (baselineResults) {
    const enrichedReducedReport = enrichReducedReportWithBaseline(reducedReport, baselineResults);
    return enrichedReducedReport.steps.map((step) => {
      const results = reportCategories.map(category => extractEnrichedResults(step, category));
      return [step.name, step.gatherMode].concat(results);
    });
  }
  return reducedReport.steps.map((step) => {
    const results = reportCategories.map(category => extractResultsValue(step.results[category]));
    // add `budgetsSymbol` to gatherMode to indicate budgets
    step.resourceCountsBudget && (step.gatherMode = step.gatherMode + ` ${budgetsSymbol}` as GatherMode);
    return [step.name, step.gatherMode].concat(results);
  });
}

function formatBytes(bytes: number, decimals = 2) {
  if (!+bytes) return '0 Bytes'

  const k = 1024
  const dm = decimals < 0 ? 0 : decimals
  const sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB']

  const i = Math.floor(Math.log(bytes) / Math.log(k))

  return `${parseFloat((bytes / Math.pow(k, i)).toFixed(dm))} ${sizes[i]}`
}


function extractTableArr(reportCategories: string[], steps: any[]): string[][] {
  const tableHead = extractTableHead(reportCategories);
  return [tableHead].concat(steps);
}

function extractTableHead(reportCategories: string[]): string[] {
  return ['Step Name', 'Gather Mode']
    .concat(reportCategories.map(c => c.split('-').map(cN => cN[0].toUpperCase() + cN.slice(1)).join(' ')));
}

function extractFractionalResultValue(fractionResults: FractionResults): string {
  const {totalWeight, numPassed, numPassableAudits} = fractionResults;
  const value = numPassed + '/' + numPassableAudits;
  return totalWeight === 0 ? 'Ã˜ ' + value : value;
}

function extractResultsValue(stepResult?: number | FractionResults): string {
  if (typeof stepResult === 'number') {
    return (stepResult * 100).toFixed(0);
  }
  return stepResult ? extractFractionalResultValue(stepResult) : '-';
}

function extractEnrichedResults(step: ReducedFlowStep, category: string): string {
  const result = extractResultsValue(step.results[category]);
  const baseline = extractResultsValue(step.baseline![category]);
  return resultWithBaselineComparison(result, baseline);
}

function resultWithBaselineComparison(result: string, baseline: string): string {
  if (result === baseline) {
    return result;
  }
  const resultNum = Number(result.replace('Ã˜ ', '').split('/')[0]);
  const baselineNum = Number(baseline.replace('Ã˜ ', '').split('/')[0]);
  const difference = baselineNum - resultNum;
  return `${result} (${difference > 0 ? '+' : ''}${difference})`;
}

function headerAlignment(reportCategories: string[]): Alignment[] {
  const reportFormats = reportCategories.map(_ => 'c');
  return ['l', 'c'].concat(reportFormats) as Alignment[];
}

