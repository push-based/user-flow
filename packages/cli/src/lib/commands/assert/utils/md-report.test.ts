import { getBudgetTable, getStepsTable } from './md-report';
import FlowResult from 'lighthouse/types/lhr/flow';
import { getReportContent } from 'test-data';
import { writeFileSync } from 'fs';
import {ReducedReport} from '../../collect/utils/report/types';
import {
  createReducedReport,
  enrichReducedReportWithBaseline,
} from '../../collect/utils/report/utils';

const lhr8 = getReportContent<any>('lhr-8.json');
const lhr9 = getReportContent<FlowResult>('lhr-9.json');
const lhr9budgets = getReportContent<FlowResult>('lhr-9-budgets.json');
const LHRREDUCEDMD = getReportContent('lhr-9_reduced.md');
const lhr9Ex2 = getReportContent<FlowResult>('lhr-9-ex-2.json');
const lhr9reduced = getReportContent<ReducedReport>('lhr-9_reduced.json');
const LHRREDUCEDCompareMD = getReportContent('lhr-9_compare.md');
const lhr9ReducedBaseline = getReportContent<ReducedReport>(
  'lhr-9_reduced-baseline.json'
);

describe('md-table', () => {
  it('should throw if version is lower than 9', () => {
    expect(lhr8['steps']).toBe(undefined);
    expect(parseFloat(lhr8.lhr.lighthouseVersion)).toBeLessThan(9);
  });

  it('should NOT throw if version is greater or equal than 9', () => {
    expect(parseFloat(lhr9.steps[0].lhr.lighthouseVersion)).toBeGreaterThan(9);
  });

  it('should generate reduced JSON format for v9 raw JSON result if createReducedReport is called', () => {
    const reducedLhr9 = createReducedReport(lhr9);
    expect(reducedLhr9).toEqual(lhr9reduced);
  });

  it('should generate reduced JSON with baseline results if enrichReducedReportWithBaseline is called', () => {
    const reducedLhr9 = createReducedReport(lhr9);
    const enrichedReducedLhr9 = enrichReducedReportWithBaseline(
      reducedLhr9,
      lhr9Ex2
    );
    expect(enrichedReducedLhr9).toEqual(lhr9ReducedBaseline);
  });

  it('should print MD table if getStepsTable is called with a reduced result', () => {
    const reducedLhr9 = createReducedReport(lhr9);
    const mdTable = getStepsTable(reducedLhr9);
    expect(mdTable).toEqual(LHRREDUCEDMD);
  });

  it('should return a Md table comparing to reports if getStepsTable is passed a baseline report', () => {
    const reducedLhr9 = createReducedReport(lhr9);
    const mdTable = getStepsTable(reducedLhr9, lhr9Ex2);
    expect(mdTable).toEqual(LHRREDUCEDCompareMD);
  });

  it('should return a Md table if getBudgetTable is passed a baseline report', () => {
    const reducedLhr9 = createReducedReport(lhr9budgets);
    const mdTable = getBudgetTable(reducedLhr9);
    expect(mdTable).toContain(
      '| Resource Type | Transfer Size | Over Budget |'
    );
    expect(mdTable).toContain('| Resource Type | Requests | Over Budget |');
    expect(mdTable).toContain(
      '|         Metric         | Measurement | Over Budget |'
    );
  });
});
