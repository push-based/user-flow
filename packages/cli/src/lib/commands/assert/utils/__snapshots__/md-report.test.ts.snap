// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`md-table should print MD table if getStepsTable is called with a reduced result 1`] = `
"| Step Name                      | Gather Mode | Performance | Accessibility | Best Practices | Seo | Pwa |
| :----------------------------- | :---------: | :---------: | :-----------: | :------------: | :-: | :-: |
| Navigation report (127.0.0.1/) | navigation  |      -      |      100      |       92       | 100 | 30  |
| Timespan report (127.0.0.1/)   |  timespan   |    10/11    |       -       |      5/7       |  -  |  -  |
| Snapshot report (127.0.0.1/)   |  snapshot   |    Ø 3/4    |     10/10     |      4/5       | 9/9 |  -  |"
`;

exports[`md-table should return a Md table comparing to reports if getStepsTable is passed a baseline report 1`] = `
"| Step Name                      | Gather Mode | Performance | Accessibility | Best Practices |   Seo    |   Pwa    |
| :----------------------------- | :---------: | :---------: | :-----------: | :------------: | :------: | :------: |
| Navigation report (127.0.0.1/) | navigation  |      -      |      100      |    92 (-7)     |   100    | 30 (+32) |
| Timespan report (127.0.0.1/)   |  timespan   |    10/11    |       -       |      5/7       |    -     |    -     |
| Snapshot report (127.0.0.1/)   |  snapshot   |    Ø 3/4    |  10/10 (-1)   |      4/5       | 9/9 (-1) |    -     |"
`;
